lib/vector-buffers/src/topology/channel/limited_queue.rs

```diff
@@ -0,0 +1,246 @@
+    /// is guaranteed that all previous values sent via `start_send` have been flushed.
+    ///
+    /// Returns `Poll::Pending` if there is more work left to do, in which case the current task is
+    /// scheduled (via `cx.waker().wake_by_ref()`) to wake up when `poll_flush` should be called again.
+    ///
+    /// If this function encounters an error, the sender should be considered to have failed
+    /// permanently, and should no longer be called.
+    pub fn poll_flush(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), SendError<T>>> {
+        match self.slot.take() {
+            None => Poll::Ready(Ok(())),
+            Some(item) => {
+                let permits_n = self.get_required_permits_for_item(&item);
+                match self.inner.limiter.poll_acquire_many(permits_n, cx) {
+                    Poll::Ready(Some(permit)) => {
+                        // We acquired enough permits to allow this send to proceed, so we bundle the
+                        // permit with the item so that when we pull it out of the queue, the capacity
+                        // is correctly adjusted.
+                        self.inner.data.push((permit, item)).expect(
+                            "queue should always have capacity when permits can be acquired",
+                        );
+
+                        // Don't forget to wake the reader since there's data to consume now. :)
+                        self.inner.read_waker.wake();
+
+                        Poll::Ready(Ok(()))
+                    }
+                    // The semaphore is closed, so the sender is closed, so the caller should not still
+                    // be calling us, so this is an error.
+                    Poll::Ready(None) => Poll::Ready(Err(SendError(item))),
+                    Poll::Pending => {
+                        // We couldn't get all the permits yet, so we have to store the item back in the
+                        // holding slot before returning.
+                        self.slot = Some(item);
+                        Poll::Pending
+                    }
+                }
+            }
+        }
+    }
+
+    /// Flush any remaining output and close this sender, if necessary.
+    ///
+    /// Returns `Poll::Ready(Ok(()))` when no buffered items remain and the sender has been successfully
+    /// closed.
+    ///
+    /// Returns `Poll::Pending` if there is more work left to do, in which case the current task is
+    /// scheduled (via `cx.waker().wake_by_ref()`) to wake up when `poll_close` should be called again.
+    ///
+    /// If this function encounters an error, the sender should be considered to have failed
+    /// permanently, and should no longer be called.
+    pub fn poll_close(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), SendError<T>>> {
+        // We can't close until any pending item is fully sent through.
+        let result = ready!(self.poll_flush(cx));
+
+        self.inner.limiter.close();
```

Comment:
I need to think a little more about this.  Given that multiple inputs may connect and disconnect to our sink, one of them calling close shouldn't close the entire channel... so we probably need to do something like move the semaphore to just `LimitedSender` and then add a `Drop` impl that only closes the semaphore once all senders are gone.

We should also add some tests, because the more I type, the more this seems obviously broken and likely to hang/stall... so we should actually have tests that can tell us if it works or not in normal shutdown conditions. 😂